cpp/src/arrow/filesystem/path_util.cc

```diff
@@ -287,6 +288,42 @@ bool IsLikelyUri(util::string_view v) {
   return ::arrow::internal::IsValidUriScheme(v.substr(0, pos));
 }
 
+struct Globber::Impl {
+  std::regex pattern_;
+
+  explicit Impl(const std::string& p) : pattern_(std::regex(PatternToRegex(p))) {}
+
+  std::string PatternToRegex(const std::string& p) {
+    std::string special_chars_ = "()[]{}+-|^$\\.&~# \t\n\r\v\f";
+    std::string transformed;
+    for (char c : p) {
+      if (c == '*') {
+        if (transformed.back() != '\\')
```

Comment:
What if `transformed` is empty here? It would seem better to structure the code a bit differently, e.g.:
```c++
auto it = p.begin();
while (it != p.end()) {
  if (*it == '\\') {
    // Escaping sequence
    if (++it == p.end()) {
      transformed += '\\';
    } else {
      transformed += *it++;
    }
    continue;
  }
  // other cases here
}

Reply:
As far as I know, `back()` in an empty string return a null char (0), so it can be compared to other chars. If you agree, I think we can keep current code structure.

Reply:
Well, you don't need to assume anything: you can read the documentation ;-)
https://en.cppreference.com/w/cpp/string/basic_string/back

> Returns reference to the last character in the string. The behavior is undefined if empty() == true.